### DISCUSSION TITLE
dynamic style attributes

### DISCUSSION THREAD

**Discussion by PierBover** [a-h/templ#958] [Q&A]
I'm having an issue where I need a dynamic `style` attribute string:

```
for _, tile := range tiles {
    <div class="tile" style={"left:" + tile.x + "px; top: " + tile.y + "px"}></div>
}
```

But `templ generate` returns an error:

```
templ parsing error: for: expected nodes, but none were found
```

The error doesn't make much sense (sorry). After trial and error I figured out the issues seems to be the dynamic `style` attribute. From the [docs](https://templ.guide/syntax-and-usage/attributes) it seems I should be able to use strings for dynamic HTML attributes:

```
<p data-testid={ testID }>Text</p>
```

What am I doing wrong?

**Comment by luxxxoor** [comment]
https://templ.guide/syntax-and-usage/css-style-management/

**Reply by PierBover** [reply]
Of course I've seen that, but I'm not interested in those features.

Is it not possible to just add a piece of dynamic string in a `style` attribute?

**Reply by amoghyermalkar123** [reply]
I tried with the style attribute but didn't find a way of making it work. But the same thing can be modeled in this way

```
type Tile struct {
	x int
	y int
}

css tileStyle(tile Tile) {
	left: { fmt.Sprintf("%dpx;", tile.x) };
	top: { fmt.Sprintf("%dpx", tile.y) };
}

templ TextAndTitle(title string, tiles []Tile) {
	<div>
		<h1 class="text-lg font-bold">{ title }</h1>
		for _, tile := range tiles {
			<div class={ tileStyle(tile) }></div>
		}
	</div>
}
```

Let me know if this helps.

**Reply by PierBover** [reply]
Thanks @amoghyermalkar123

I gotta say, it's quite an unintuitive design decision to be using dynamic styles via creating a CSS class. Pretty much any HTML templating language since the 90s has been able to just create dynamic inline styles.

**Reply by pharrisee** [reply]
It seems that the reason for this is to ensure sanitised CSS only in templ (at least that's my reading of the docs), but, how about abusing the [templ.Attributes](https://templ.guide/syntax-and-usage/attributes/#spread-attributes) spreading to achieve it?

```go
templ Index() {
	<h1 { styles("blue")... }>Templ Components</h1>
}

func styles(bgColor string) templ.Attributes {
	return templ.Attributes{"style": fmt.Sprintf("color:red;background-color:%s", bgColor)}
}
```

btw, the error I see when trying to have a calculated `style` attribute is slightly different to the error you see:

```
invalid style attribute: style attributes cannot be a templ expression
```

EDIT:

Just also tried this and it seems to work as well:

```go
templ Index() {
	<h1 { templ.Attributes{"style":"color:"+"green"}... }>Templ Components</h1>
}
```

I can only presume that this works as the attributes are sanitised in some way...

**Reply by PierBover** [reply]
Thanks @pharrisee the attributes solution works well. Not sure why CSS would need to be sanitized to be honest.

Another point regarding the `css` functions is that creating `<style>` tags in the `<body>` tag is invalid HTML5. Try it out in the [official HTML validator](https://validator.w3.org/#validate_by_input).

Finally, creating new classes for every variation of some values can be highly inefficient compared to inline styles. In my benchmark I got approx 3x performance with inline styles.

**Reply by PierBover** [reply]
For reference this is what ended up working:

```go
templ Tiles(tiles []Tile) {
	for _, tile := range tiles {
		<div class="tile" {templ.Attributes{"style":"top:" + tile.x + "px;left:" + tile.y + "px;"}...}></div>
	}
}
```

**Reply by roborobs1023** [reply]
> For reference this is what ended up working:
> 
> ```go
> templ Tiles(tiles []Tile) {
> 	for _, tile := range tiles {
> 		<div class="tile" {templ.Attributes{"style":"top:" + tile.x + "px;left:" + tile.y + "px;"}...}></div>
> 	}
> }
> ```

Seems as this is similar to the way react does style attributes but this allows you to insert go variables into the code.
